Factor out database specific dependencies into separate packages

basti1302 opened the pull request:
**DO NOT MERGE THIS YET**

This is a first draft of a solution to
https://github.com/jtdaugherty/dbmigrations/issues/28

All MySQL specific parts have been removed from this package and
bundled into a new package called dbmigrations-mysql,
The main goal here is to not force people who want to use dbmigrations
with PostgreSQL or any other database backend to install MySQL
components on their systems.

There's probably still some more stuff to do but I wanted to check with you if this goes into the right direction.

Here's what I did so far:

* The `Configuration` as well as the `AppState` can now hold either a dbType and dbConnection string (as before) or a fully instantiated backend. If the appState has dbType and dbConnection, all works as before. If there is already a instantiated backend instead of the two strings, this is simply used in `CommandUtils#withBackend`.
* All MySQL specific code has been moved to a new package `dbmigrations-mysql` that is just a thin wrapper around `dbmigrations`. Currently, this new package lives here: https://github.com/basti1302/dbmigrations-mysql
* I have not yet removed any PostgreSQL or SQLite specific code.

I was able to upgrade a MySQL database with the executable `moo-mysql` from the new package and upgrading a PostgreSQL db with `moo` also still works.

If you agree with the general direction here so far, I would like to discuss a few things to get consensus about how to proceed with this:
* Do you really want to factor out PostgreSQL also? What about SQLite? Neither of them come with the problems (build time dependencies) that lead to this refactoring, so I think it would be a valid course of action to just leave support for them in the main package. OTOH, this feels a bit asymetrical and arbitrary. What do you think?
* I did not take care of the tests at all until now. I had one test failure on a clean checkout and I'm not sure what to do about that. Here's the [log](https://gist.github.com/basti1302/911ede318501d2dd5b2257bcd589afd3). Do the tests work for you on current master?
* Where will the dbmigrations-mysql repository go in the end and who will publish it etc.? My idea right now would be that we transfer this repo to your account at some point. I'm actually not too keen on maintaining it, since the reason for all this is that I do _not_ use MySQL ;-)

Let me know what you think.

Cheers

  Bastian

jtdaugherty commented:
This is fantastic - thank you! I should have some time to take a close look at this today and respond in detail.

jtdaugherty commented:
> Do you really want to factor out PostgreSQL also? What about SQLite? Neither of them come with the problems (build time dependencies) that lead to this refactoring, so I think it would be a valid course of action to just leave support for them in the main package. OTOH, this feels a bit asymetrical and arbitrary. What do you think?

I would definitely like to move all backend-specific functionality out of the core library into `dbmigrations-<backend>` packages. This was a problem before MySQL support was added and frankly it's just not necessary to have them all in the same package.

> I did not take care of the tests at all until now. I had one test failure on a clean checkout and I'm not sure what to do about that. Here's the log. Do the tests work for you on current master?

I'll look into that failure; I'll have to do some work to get a testing environment set up, but since this was present on a clean checkout I think you can safely ignore it for now. Thanks for running the tests!

> Where will the dbmigrations-mysql repository go in the end and who will publish it etc.? My idea right now would be that we transfer this repo to your account at some point. I'm actually not too keen on maintaining it, since the reason for all this is that I do not use MySQL ;-)

I'm happy to accept full responsibility for it (Hackage, github, maintenance), although I would like you to get credit for doing it somehow - either by a mention in the README (and certainly in the CHANGELOG) or in the package description metadata.

As for the specific changes:

Overall it looks good to me. In particular I see some of these changes as transitional; these changes make me want the core library to be parameterized only by a configuration and a backend, rather than taking the bits necessary to create a backend. That would also entail making the application state only store a `Backend`, and that would simplify a lot of configuration handling where it currently now branches on `Left` and `Right` to figure out whether a backend is already present. Furthermore this would make it possible to get rid of the database "type" setting since that would be implicit in the `Backend`.

As part of that I'd also like to make it possible to hide all the details of connecting to each backend (such as command-line arguments) in each respective backend-specific `moo` executable. That way, backends that don't support connection strings don't need a hack to conform to the library interface that expects one. That would also probably require the core library to expose an API for adding extra command-line options to the set currently handled by the argument processing. But all that can come later.

Everything I'm describing here is just future directions stuff, not a request for you to do more work on the patch. (Certainly if you wanted to, I'd welcome it - but this is a much-needed first step and I'd be happy to accept this patch by itself without the other stuff I mentioned above.)

jtdaugherty commented:
Regarding the failing test you mentioned, I actually forgot that it's expected:
https://github.com/jtdaugherty/dbmigrations/commit/a50490957dcecd61883fd61ae21633a5d08b6668

jtdaugherty commented:
(This was a reason I had major reservations about even supporting MySQL, even though folks who use it are probably okay with this behavior.)

basti1302 commented:
Somehow I was in the mood and did some more things, going in the direction you outlined in your comments. All three supported database types have been factored out into separate packages:

* https://github.com/basti1302/dbmigrations-postgresql
* https://github.com/basti1302/dbmigrations-mysql
* https://github.com/basti1302/dbmigrations-sqlite

There is now an explicit distinction between what the core library parses as a configuration (this type contains a connection string but no Backend) and what the executable from the wrapper package passes to the core lib to execute an actual command (contains no connection string, but an actual Backend).

I also updated some docs and fixed the failing tests.

I did not work on the use case you mentioned for backend types that do not have a notion of a connection string at all that.

The BackendTest has been moved to the three new, backend-specific packages. There is some duplication (BackendTest.hs is nearly the same and TestDriver.hs is quite similar in all three packages) but I think this might be okay-ish.

I have some woes with the BackendTest for MySQL, for which I created an [issue](https://github.com/basti1302/dbmigrations-mysql/issues/1) in the new package for MySQL. I won't investigate any further there, maybe you can figure that one out more easily or the original author for the MySQL support.

I would like you to give all this a thorough review, I think this is ready to merge or close enough. Give me a ping when you want me to squash the commits in this PR before merging.

Kind regards

  Bastian

PS/side note: There is also a small  [syntactical issue](https://github.com/jtdaugherty/dbmigrations/pull/32/commits/39106b4b074e7edd6c877583f6d54afa3828b983#diff-c974e3fffb8fa5dbb6b733db2c903d25R95) I'd like to draw your attention to (I'm quite new to Haskell, so bear with me).

jtdaugherty commented on src/Moo/CommandUtils.hs:
```diff
@@ -88,30 +87,15 @@ lookupMigration storeData name = do
+  -- unfortunately I'm not very familiar with the Reader monad :-/
+  backendIO <- asks _appBackend
+  let backend = return backendIO
   liftIO $ bracket backend disconnectBackend act
```
Since this code is pulling the `backend` out of `AppT` and then using it in `IO`, I think the way you've written it makes sense. But since the "before" action of `bracket` isn't really doing anything, I'd probably write it as
```
liftIO $ bracket (return ()) (const $ disconnectBackend backendIO) act
```
since when I read `bracket` I expect the "before" action to be doing something critical to the "after" action.  Even more, since we aren't using bracket in the usual way here, another way to write this is just
```
liftIO $ act `finally` (disconnectBackend backendIO)
```

jtdaugherty commented:
> Somehow I was in the mood and did some more things, going in the direction you outlined in your comments. All three supported database types have been factored out into separate packages:
> https://github.com/basti1302/dbmigrations-postgresql
> https://github.com/basti1302/dbmigrations-mysql
> https://github.com/basti1302/dbmigrations-sqlite

Fantastic!

> There is now an explicit distinction between what the core library parses as a configuration (this type contains a connection string but no Backend) and what the executable from the wrapper package passes to the core lib to execute an actual command (contains no connection string, but an actual Backend).

That is what I had in mind; thank you!

> The BackendTest has been moved to the three new, backend-specific packages. There is some duplication (BackendTest.hs is nearly the same and TestDriver.hs is quite similar in all three packages) but I think this might be okay-ish.

I would very much like to keep the core test suite in the library so that it is possible to know whether each backend passes muster without worrying about test code drift or duplication. I don't want to be in a situation where one backend worked fine with some version of `dbmigrations` but we didn't see the others failing because their test suites hadn't been updated. I'd like to think of the tests as a "conformance suite" provided by the library to make it easy for backend implementors to know that they got something right. The test programs in each backend package could be like the `moo-*` executables: just parameterizing the test suite with a `Backend`. This also has the benefit that there'll be one copy of the set of migrations used test. (And we'll deal with the MySQL known-failing test later.)

> I have some woes with the BackendTest for MySQL, for which I created an issue in the new package for MySQL. I won't investigate any further there, maybe you can figure that one out more easily or the original author for the MySQL support.

Okay. Thanks for reporting that! I'll look into it later. (And if I get the motivation, I'll move away from `mysql` anyway; its maintainer has been unresponsive for some time.)

> I would like you to give all this a thorough review, I think this is ready to merge or close enough. Give me a ping when you want me to squash the commits in this PR before merging.

If we can resolve the test suite issue I mentioned above, I think I'd be okay with merging this and taking over the other repos you created. However, please do not squash your commits!  There are few enough and they're clean enough that they're more useful as-is than as one monster patch.

> PS/side note: There is also a small syntactical issue I'd like to draw your attention to (I'm quite new to Haskell, so bear with me).

I believe I commented on what you're referring to here.

> PPS: I also did not add my name in the readme or changelog or anywhere. You are of course welcome to give me credit however you see fit :-)

Okay, will do.

Thanks for your work on this!

basti1302 commented:
I incorporated your latest suggestions. The BackendEnd test has been moved back to the core package. I believe it needs to be in the library, not the test suite so that we can export it. This has the consequence that the lib now depends on HUnit.

There is still minimal duplication between the postgres and the sqlite tests, as the `instance BackendConnection PostgreSQLBackendConnection` and `instance BackendConnection SQLiteBackendConnection` are basically the same expect the implementation of `getDatabaseType`.
Maybe that can be refactored further to a common HDBC backend thing in core, but anyways, I think I'd like to stop at this point :-)

Would you like to take over now? From my point of view, that would entail merging this PR and moving the three new repos to your account.

Thanks for your support and suggestions.

jtdaugherty commented:
> I incorporated your latest suggestions. The BackendEnd test has been moved back to the core package. I believe it needs to be in the library, not the test suite so that we can export it. This has the consequence that the lib now depends on HUnit.

Yeah, that's what I expected, and I think it's okay since the tests are part of the library interface now.

> There is still minimal duplication between the postgres and the sqlite tests, as the instance BackendConnection PostgreSQLBackendConnection and instance BackendConnection SQLiteBackendConnection are basically the same expect the implementation of getDatabaseType.
Maybe that can be refactored further to a common HDBC backend thing in core, but anyways, I think I'd like to stop at this point :-)

I'm happy to take over if anything needs to be done there. But I have one question: what is the `DatabaseType` enumeration for? I don't see any places that use it, and I'd like to either make it more generic (contain a string) so that the core library doesn't have to have *any* knowledge about the backends that exist.

> Would you like to take over now? From my point of view, that would entail merging this PR and moving the three new repos to your account.

I'm happy to take over once the enumeration bit is clearer.

> Thanks for your support and suggestions.

You're welcome; thanks for doing this work!

basti1302 commented:
Regarding the DatabaseType enumeration: This is currently only used in [BackendTest](https://github.com/jtdaugherty/dbmigrations/pull/32/files#diff-5d6d9f49848e68da16a41daa9354c367R78) to decide which tests to run. MySQL gets a reduced test suite because of its transactional DDL insufficencies. I went for an enumeration instead of a string because I thought it might be useful later to have this. On second thought, you're right, it's clearly YAGNI and introduces an unnecessary coupling between backend wrapper packages and core. A string would be sufficient and better. Actually, I think replacing the DB type with a boolean flag `supportsTransactionalDDL` would be even better here.

Do you want me to change it or do you want to go ahead with it?

jtdaugherty commented:
I would actually like for that test to continue failing; I want to make it so that it's a noisy but expected failure (unlike an unexpected failure now) so anyone running those gets the point. I want people to be very afraid of using MySQL with this software. :)

But I'm happy to carry on with the work from here - I just wanted to be sure I understood the enumeration's role in things. Thanks!

jtdaugherty commented:
I'm now ready to assume ownership of those other repos.

basti1302 commented:
> I would actually like for that test to continue failing; I want to make it so that it's a noisy but expected failure (unlike an unexpected failure now) so anyone running those gets the point. I want people to be very afraid of using MySQL with this software. :)

Sure, it's your call, but I would strongly advise against this. As I tried to explain in https://github.com/basti1302/dbmigrations-mysql/issues/1 the failing test causes most of the test cases in `BackendTest` to **not be run at all** for MySQL. Since `applyMigrationFailure` is the fourth of eight test cases and BackendTest stops when it encouters one failing test, the remaining four test cases never get run. This means, if you break something in the code under test, you'll never notice by running the tests cause all you see is the expected failure that is always there.

Pluswhich, you already have a prominent warning in MOO.TXT. If you feel the need, add more warnings in the README, dbmigrations.cabal, dbmigrations-mysql.cabal or even print an annoying warning at every start of `moo-mysql`. 

People using dbmigrations-mysql will probably not run the tests at all anyway, so having a failing test wouldn't help, it also does not bring the point across. Even if a user ran the test they'd probably just shrug and think "hm, slightly shitty package with a failing unit test on a clean checkout".

Sorry for the rant :-) but IMHO leaving a know test failure in your repo really never ever is a benefit.

PS: I transferred the other repos.

jtdaugherty commented:
> Sure, it's your call, but I would strongly advise against this. As I tried to explain in jtdaugherty/dbmigrations-mysql#1 the failing test causes most of the test cases in BackendTest to not be run at all for MySQL. Since applyMigrationFailure is the fourth of eight test cases and BackendTest stops when it encouters one failing test, the remaining four test cases never get run. This means, if you break something in the code under test, you'll never notice by running the tests cause all you see is the expected failure that is always there.

Agreed! That's why I mentioned making it an "expected failure" - although I don't know how to do that given the current framework. Another option would be to make it so that the test suite doesn't fail on the first failure, but runs all the tests and then exists nonzero if any failed. (I'd prefer that approach anyway for exactly the reason you described.)
